### PR TITLE
Limit layer height to `max_layer_height` if defined when combining infills 

### LINKED ISSUE #2735: "Infill combination" feature ignores layer height limitations from "printer settings" section
### OrcaSlicer Version

1.8.0-rc2

### OS version

Windows 10, Windows 11

### Additional system information

_No response_

### Printer

Kingroon KP3S (custom preset)

### How to reproduce

1. Go to "Extruder" tab in 'Printer settings"
2. Set max layer height limit less than [layer_height * n]. For example, I tried it with 0.4 max layer height limit and 0.125 layer height
3. Enable "Infill combination" feature (process settings - strength - advanced)
image

### Actual results

Maximum layer height becomes greater than layer height limit and proportional to current layer height

### Expected results

Maximum layer height should be less than maximum layer height limit

### Project file & Debug log uploads

[Layer Height Test.zip](https://github.com/SoftFever/OrcaSlicer/files/13348960/Layer.Height.Test.zip)

### Checklist of files to include

- [ ] Log file
- [X] Project file

### PR DESCRIPTION
**Pull request by Noisyfox**
This fixes #2735

### CODE REVIEW

File: src/libslic3r/PrintObject.cpp [hunk 1]
```diff
@@ -3306,6 +3306,13 @@ void PrintObject::discover_horizontal_shells()
 // fill_surfaces but we only turn them into VOID surfaces, thus preserving the boundaries.
 void PrintObject::combine_infill()
 {
+    auto max_layer_height = [this](int idx_nozzle) {
+        auto    &print_config = this->print()->config();
+        coordf_t mh           = print_config.max_layer_height.get_at(idx_nozzle - 1);
```
Comment (SoftFever) [inline src/libslic3r/PrintObject.cpp:3311]: should we use `idx_nozzle` instead here?

Comment (Noisyfox) [reply src/libslic3r/PrintObject.cpp:3311]: Ah that's right.

**Comment by SoftFever** [comment]
I'm a bit hesitant to change the current behavior. Many users, myself included, are used to the existing limitation for combined infill, which is determined by the nozzle diameter. In a typical scenario (0.4 mm nozzle and a 0.2 layer height), using a 0.4 mm layer height for combined infills is pretty practical.

However, the proposed changes would force users to increase the maximum layer height in their machine settings, which is not ideal. For example, with a 0.4 mm nozzle, we typically set the max layer height to 0.32 mm, not 0.4 mm.
Combined infill is kinda of an exception

**Comment by Noisyfox** [comment]
Yeah I'm kinda having the same concerns...

**Comment by igiannakas** [comment]
> I'm a bit hesitant to change the current behavior. Many users, myself included, are used to the existing limitation for combined infill, which is determined by the nozzle diameter. In a typical scenario (0.4 mm nozzle and a 0.2 layer height), using a 0.4 mm layer height for combined infills is pretty practical.
> 
> However, the proposed changes would force users to increase the maximum layer height in their machine settings, which is not ideal. For example, with a 0.4 mm nozzle, we typically set the max layer height to 0.32 mm, not 0.4 mm. Combined infill is kinda of an exception

I agree - I wouldn't change this behaviour. When combining infill you typically want it to go as fast as possible, which means that you'd want as many layers as possible combined against the nozzle size. 0.2 layer height is a bit of a problem spot as it combines to a 0.4 layer height infill that is comparatively weak due to lack of "squish", compared to a 0.16 layer height that combines to 0.32 which is fine.

Cura has an option to combine every X layers, which may be useful for smaller layer heights (eg. 0.12 over 2 layers instead of 3 which is the Orca behaviour), but again that would have limited value as it doesn't solve the 0.4 layer height "problem" above.

**Comment by igiannakas** [comment]
An alternative approach could be to allow a max layer height specifically for infill when layer height is combined, shown as an option on the strength tab. Allow for %'s and absolute value so this can be set independently of the min/max layer height of the extruder (which also affects variable layer height limits). This can be set to default at 80% which is the recommended max layer height in order to have decent layer adhesion. 

That would mean that a 0.1 layer height model would combine infill to a good maximum of 3 layers (0.3mm) instead of 4 (0.4) that would result in higher strength models. 

Conversely this would also allow flexibility for 0.2 layer height models to choose to override that setting and live with a weaker infill without affecting other things, like variable layer height limits which are tied to the extruder values.

**Comment by SoftFever** [comment]
Adding a new option just for this is an overkill @_@

**Comment by stew675** [comment]
> Adding a new option just for this is an overkill @_@

What about putting it into the Extruder section for the machine?  Most everything there is set once and forget anyway.  That way it's not cluttering up the main work-flow interface but still provides the flexibility that people want.

Personally though, I'd opt for a value like "maximum number of layers to combine" so that way it's independent of nozzle size/layer-height, with the upper limit being as it currently is.

**Comment by karlc28** [comment]
Every other slicer has an option to set the amount of layers to combine.
It seems strange that this would be emitted from orca

the ideal solution would be when selecting infill combination that you can specify how many layers 

if im printing at 0.1 layer height I want to be ale to have infill at 0.2 or 0.3 layer height atm orca sets it at 0.4 which is ridiculous

**Comment by igiannakas** [comment]
> . I cannot use infill combination in my prints because of this. I have max layer height set to 0.2mm on my 0.4mm nozzle and infill combination still wants to print 0.4mm high infills.

Combination infill prints multiples of your layer height. If you have LH at 0.2 it can only do 0.4 lines (combine every two layers). If this option was implemented it wouldn’t combine your infill at all.

**Comment by vgdh** [comment]
0,4mm is with 0.4mm nozzle is to high for infill.
For this case may be the solution is to use not equal layer height for infill. 
For example with walls 0.2mm height use infill with layers height 0.1and 0.2 alternatively.

**Comment by Gamerou** [comment]
> > . I cannot use infill combination in my prints because of this. I have max layer height set to 0.2mm on my 0.4mm nozzle and infill combination still wants to print 0.4mm high infills.
> 
> 
> 
> Combination infill prints multiples of your layer height. If you have LH at 0.2 it can only do 0.4 lines (combine every two layers). If this option was implemented it wouldn’t combine your infill at all. 
> 
> 
> 
> 

The speed aspect is crucial here. With a normal layer height of 0.2mm, it would be possible to combine infill layers to 0.3mm or 0.32mm, which would allow for faster prints. It's not about combining entire layers, but that's beside the point. Every other slicer has this option. Why doesn’t Orca?

**Pull request closed without merging**